### ISSUE TITLE
Add allowance, approve and transferFrom to NEP-17

### ISSUE THREAD

**Issue by lock9** [neo-project/neo#2208]
**Summary or problem description**
We are building a DeFi Dapp and we need the allowance feature for it to work 'properly'. Without the allowance feature, we will depend on manual intervention from users - doable, but definitely not the best case.

**Do you have any solution you want to propose?**
Add Allowance, Approve, and TransferFrom methods to Nep-17 (including NEO and GAS)

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Other: NEP-17

**Comment by erikzhang** [comment]
Use `transfer` and `onPayment`.

**Comment by lock9** [comment]
Hmm, I don't think this will work. Problem:
I have 100 tokens on my balance. I want to put this tokens for sale on two different Dexes at the same time. How can I do that without allowance?
It is not possible, because if I use ```transfer```, the balance is removed from my account. If I use allowance, both SC will have the allowance to use the 100 tokens, meaning I can put my tokens to use on different dexes, at the same time.

**Comment by igormcoelho** [comment]
I think this is a recurrent problem... shouldn't we propose another NEP, that extends NEP-5/17 with these two methods? After that, we may discuss if native NEO/GAS should be NEP-17 or NEP-17+... what do you think about this direction?

**Comment by erikzhang** [comment]
I don't think it has many meaningful scenarios. In fact, most situations can be solved with `transfer` + `onPayment`. This is more simple.

**Comment by lock9** [comment]
> I think this is a recurrent problem... shouldn't we propose another NEP, that extends NEP-5/17 with these two methods? After that, we may discuss if native NEO/GAS should be NEP-17 or NEP-17+... what do you think about this direction?

I think Allowance is needed for everything that can be owned, fungible, or not.



> I don't think it has many meaningful scenarios. In fact, most situations can be solved with `transfer` + `onPayment`. This is more simple.

How can I make my tokens available to two different smart contracts? If I use 'transfer', I will lose ownership of my assets. I need a way to "put my bicycle for sale at multiple places, and still be able to use it".  If I use transfer, I can't use these tokens. I don't think there is a replacement for the allowance method.
How can I implement this @erikzhang ? Maybe I'm missing something...

**Comment by bibicoco2015** [comment]
Can you please guide me on how to prevent automatic emails from being sent back to your inbox.![image](https://user-images.githubusercontent.com/45157636/104407025-4ddfd400-5593-11eb-892d-7bd036cbe8a5.png)

**Comment by erikzhang** [comment]
> How can I make my tokens available to two different smart contracts?

From a developer's perspective, why would you allow the uses to make their tokens available to different contracts? I think that developers would prefer users to choose his contract instead of choosing other contracts.

From a user's perspective, I agree that users may want to be available in both contracts. But this increases the difficulty of operation and the chance of errors. Users may be confused about where the tokens are going.

From a system perspective, this will generate a lot of uncashable checks. If one contract transfers these tokens, other contracts will fail. Generating a lot of failed transactions is not a good design.

To sum up, I think it is not worth adding these interfaces.

**Comment by lock9** [comment]
> From a developer's perspective, why would you allow the uses to make their tokens available to different contracts? I think that developers would prefer users to choose his contract instead of choosing other contracts.

There are multiple reasons for that. One can be that all these multiple contracts belong to the same dApp. Another reason can be that the user has different price positions on different markets and wants to allow his tokens to be sold when certain criteria are met.  

> From a user's perspective, I agree that users may want to be available in both contracts. But this increases the difficulty of operation and the chance of errors. Users may be confused about where the tokens are going.

They will want to use multiple contracts because they want to increase their profits. I agree that this is an advanced feature and it enables multiple new use cases. 
Regarding the chance of errors, I think that they will likely occur, but maybe we should leave this out of the discussion. There are chances of errors everywhere. There are people who put their private keys on webpages they got from strangers in Discord. 


> From a system perspective, this will generate a lot of uncashable checks. If one contract transfers these tokens, other contracts will fail. Generating a lot of failed transactions is not a good design.

I can't argue with that. If you say, I believe. I have a question though:  if a transaction has a higher network fee attached, can we presume what transaction will be persisted? This may lead to people competing to send a transaction with a higher network fee. If this is good or bad, I can't say. I noticed that people on the Ethereum network tend to dislike paying fees to miners, but their fees are much more expensive.

> To sum up, I think it is not worth adding these interfaces.
Well, the system part is something that I really can't argue against. I really wish we didn't have to use a wrapped Neo, but I agree that if we add this to the core classes, it will affect all accounts and if we add it to the wrapped version, it will only affect those using the wrapped version. 

I think that the solution is to have another NEP, as Igor described, and instead of adding it to Neo and Gas, I should use build a wrapper instead.
